File: frontend/src/schemas/portfolio.ts
import { z } from 'zod'

export const StatItemSchema = z.object({
  value: z.string(),
  accent: z.string(),
  label: z.string(),
})
export type StatItem = z.infer<typeof StatItemSchema>

export const DEFAULT_STATS: StatItem[] = [
  { value: '500', accent: '+', label: 'Sessions completed' },
  { value: '10', accent: '+', label: 'Years of experience' },
  { value: '5', accent: ' ★', label: 'Average client rating' },
  { value: '48', accent: 'h', label: 'Photo delivery time' },
]

// Public
export const HeroPhotoSchema = z.object({
  id: z.string().uuid(),
  image_url: z.string().url(),
  position: z.number(),
})
export type HeroPhoto = z.infer<typeof HeroPhotoSchema>

export const CategorySchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  slug: z.string(),
  cover_url: z.string().url(),
  position: z.number(),
})
export type Category = z.infer<typeof CategorySchema>

export const PhotoSchema = z.object({
  id: z.string().uuid(),
  image_url: z.string().url(),
  position: z.number(),
})
export type Photo = z.infer<typeof PhotoSchema>

export const CategoryWithPhotosSchema = CategorySchema.extend({
  photos: z.array(PhotoSchema).default([]),
})
export type CategoryWithPhotos = z.infer<typeof CategoryWithPhotosSchema>

export const PublicSettingsSchema = z.object({
  tagline: z.string().nullable().optional(),
  bio: z.string().nullable().optional(),
  instagram_url: z.string().nullable().optional(),
  facebook_url: z.string().nullable().optional(),
  contact_headline: z.string().nullable().optional(),
  admin_name: z.string().nullable().optional(),
  admin_avatar_url: z.string().nullable().optional(),
  meta_title: z.string().nullable().optional(),
  meta_description: z.string().nullable().optional(),
  og_image_url: z.string().nullable().optional(),
  stats: z.array(StatItemSchema).optional().default([]),
})
export type PublicSettings = z.infer<typeof PublicSettingsSchema>

export const AboutSettingsSchema = z.object({
  tagline: z.string().nullable().optional(),
  bio: z.string().nullable().optional(),
  instagram_url: z.string().nullable().optional(),
  facebook_url: z.string().nullable().optional(),
  contact_headline: z.string().nullable().optional(),
  contact_email: z.string().nullable().optional(),
  meta_title: z.string().nullable().optional(),
  meta_description: z.string().nullable().optional(),
  og_image_url: z.string().nullable().optional(),
  stats: z.array(StatItemSchema).optional().default([]),
})
export type AboutSettings = z.infer<typeof AboutSettingsSchema>

// Contact form
export const ContactFormSchema = z.object({
  name: z.string().min(1, 'Name is required').max(200),
  email: z.string().email('Invalid email address').max(254),
  phone: z.string().max(30).optional(),
  message: z.string().min(1, 'Message is required').max(5000),
})
export type ContactForm = z.infer<typeof ContactFormSchema>

// Admin operations
export const PositionItemSchema = z.object({
  id: z.string().uuid(),
  position: z.number().int().positive(),
})
export type PositionItem = z.infer<typeof PositionItemSchema>

File: backend/schemas/portfolio.py
from __future__ import annotations
import json
from typing import Optional
from uuid import UUID
from datetime import datetime
from pydantic import BaseModel, field_validator, EmailStr
import re


class HeroPhotoOut(BaseModel):
    id: UUID
    image_url: str
    position: int
    model_config = {"from_attributes": True}


class PositionItem(BaseModel):
    id: UUID
    position: int


class CategoryOut(BaseModel):
    id: UUID
    name: str
    slug: str
    cover_url: str
    position: int
    model_config = {"from_attributes": True}


class PhotoOut(BaseModel):
    id: UUID
    image_url: str
    position: int
    model_config = {"from_attributes": True}


class CategoryWithPhotosOut(CategoryOut):
    photos: list[PhotoOut] = []


class ContactSubmissionCreate(BaseModel):
    name: str
    email: EmailStr
    phone: Optional[str] = None
    message: str

    @field_validator("name")
    @classmethod
    def name_length(cls, v: str) -> str:
        v = v.strip()
        if not v or len(v) > 200:
            raise ValueError("name must be 1–200 characters")
        return v

    @field_validator("phone")
    @classmethod
    def phone_length(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        v = v.strip()
        if len(v) > 30:
            raise ValueError("phone must be 30 characters or fewer")
        return v or None

    @field_validator("message")
    @classmethod
    def message_length(cls, v: str) -> str:
        v = v.strip()
        if not v or len(v) > 5000:
            raise ValueError("message must be 1–5000 characters")
        return v


class ContactSubmissionOut(BaseModel):
    id: UUID
    name: str
    email: str
    phone: Optional[str] = None
    message: str
    created_at: datetime
    model_config = {"from_attributes": True}


DEFAULT_STATS = [
    {"value": "500", "accent": "+", "label": "Sessions completed"},
    {"value": "10", "accent": "+", "label": "Years of experience"},
    {"value": "5", "accent": " ★", "label": "Average client rating"},
    {"value": "48", "accent": "h", "label": "Photo delivery time"},
]


class StatItem(BaseModel):
    value: str
    accent: str
    label: str


class PublicSettingsOut(BaseModel):
    tagline: Optional[str] = None
    bio: Optional[str] = None
    instagram_url: Optional[str] = None
    facebook_url: Optional[str] = None
    contact_headline: Optional[str] = None
    admin_name: Optional[str] = None
    admin_avatar_url: Optional[str] = None
    stats: list[StatItem] = []


class AboutSettingsOut(BaseModel):
    tagline: Optional[str] = None
    bio: Optional[str] = None
    instagram_url: Optional[str] = None
    facebook_url: Optional[str] = None
    contact_headline: Optional[str] = None
    contact_email: Optional[str] = None
    meta_title: Optional[str] = None
    meta_description: Optional[str] = None
    og_image_url: Optional[str] = None
    stats: list[StatItem] = []


class AboutSettingsUpdate(BaseModel):
    tagline: Optional[str] = None
    bio: Optional[str] = None
    instagram_url: Optional[str] = None
    facebook_url: Optional[str] = None
    contact_headline: Optional[str] = None
    contact_email: Optional[str] = None
    meta_title: Optional[str] = None
    meta_description: Optional[str] = None
    stats: Optional[list[StatItem]] = None

    @field_validator("contact_email")
    @classmethod
    def valid_email(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        if "@" not in v or len(v) > 254:
            raise ValueError("contact_email must be a valid email address.")
        return v

    @field_validator("instagram_url", "facebook_url")
    @classmethod
    def valid_url(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        if not re.match(r'^https?://', v):
            raise ValueError("must be a valid HTTP/HTTPS URL.")
        return v
